hub/core/storage/lru_cache.py

```diff
@@ -0,0 +1,163 @@
+from collections import OrderedDict
+from hub.core.storage.provider import StorageProvider
+from typing import Set
+
+# TODO use lock for multiprocessing
+class LRUCache(StorageProvider):
+    """LRU Cache that uses StorageProvider for caching"""
+
+    def __init__(
+        self,
+        cache_storage: StorageProvider,
+        next_storage: StorageProvider,
+        cache_size: int,
+    ):
+        """Initializes the LRUCache. This cache can be chained with
+
+        Args:
+            cache_storage (StorageProvider): The storage being used as the caching layer of the cache the.
+                This should be a base provider such as MemoryProvider, LocalProvider or S3Provider but not another LRUCache.
+            next_storage (StorageProvider): The next storage layer of the cache.
+                This can either be a base provider (i.e. it is the final storage) or another LRUCache (i.e. in case of chained cache).
+                While reading data, all misses from cache would be retrieved from here.
+                While writing data, the data will be written to the next_storage when cache_storage is full or flush is called.
+            cache_size (int): The total space that can be used from the cache_storage in bytes.
+                This number may be less than the actual space available on the cache.
+                Setting it to a higher value than actually available space may lead to unexpected behaviors.
+        """
+        self.next_storage = next_storage
+        self.cache_storage = cache_storage
+        self.cache_size = cache_size
+
+        # tracks keys in lru order, stores size of value, only keys present in this exist in cache
+        self.lru_sizes: OrderedDict[str, int] = OrderedDict()
+        self.dirty_keys: Set[str] = set()  # keys present in cache but not next_storage
+        self.cache_used = 0
+
+    def flush(self):
+        """Writes data from cache_storage to next_storage. Only the dirty keys are written.
+        This is a cascading function and leads to data being written to the final storage in case of a chained cache.
+        """
+        for key in self.dirty_keys:
+            self.next_storage[key] = self.cache_storage[key]
+        self.dirty_keys.clear()
+
+        self.next_storage.flush()
+
+    def __getitem__(self, path: str):
+        """If item is in cache_storage, retrieves from there and returns.
+        If item isn't in cache_storage, retrieves from next storage, stores in cache_storage (if possible) and returns.
+
+        Args:
+            path (str): the path relative to the root of the underlying storage.
+
+        Returns:
+            bytes: The bytes of the object present at the path
+        """
+        if path in self.lru_sizes:
+            self.lru_sizes.move_to_end(path)  # refresh position for LRU
+            return self.cache_storage[path]
+        else:
+            result = self.next_storage[path]  # fetch from storage
```

Comment:
What happens if it's not in the next storage either? If we're expecting this to throw, we should probably add a little bit to the docstring indicating that it might occur